Swift bindings generator with simple parser and emitter

kotlarmilos opened the pull request:
## Description

This PR introduces a toolset for generating C# bindings from Swift interface files. This is the MVP aiming to support the generation of C# bindings for [CryptoKit dev templates](https://github.com/dotnet/runtime/issues/95636).

The tooling consists of the following components:
- **SwiftBindings**: A command-line interface with Swift ABI parser, marshalling types, and string-based C# emitter.
- **SwiftRuntimeLibrary**: A library providing projections of common Swift types.

The parser and emitter are modular and implement `ISwiftParser` and `ICSharpEmitter` interfaces. The Swift ABI parser can be replaced with a swiftinterface parser, while the string-based emitter can be replaced with an object model-based emitter.

**Workflow:**
1. Parse Swift ABI files to aggregate public ABI.
2. Generate marshalling information if needed.
3. Emit C# source code using a string-based emitter.

The unit tests verify the XML layout of the type database and various P/Invoke cases. Additionally, CryptoKit dev templates with unsafe (buffer) pointers are implemented.

kotlarmilos committed:
Add SwiftRuntimeLibrary project

The SwiftRuntimeLibrary provides runtime marshaling support and projections of common Swift types.

kotlarmilos committed:
Add arm64 CI job

rolfbjarne commented on src/SwiftRuntimeLibrary/src/ISwiftObject.cs:
```diff
@@ -0,0 +1,13 @@
+using System;
+using System.Runtime.InteropServices;
+
+namespace SwiftRuntimeLibrary
```
What about this instead:

```suggestion
namespace Swift.Runtime
```

kotlarmilos replied:
Sounds good. Is it compatible with ObjC namespace naming?

rolfbjarne replied:
Yes, I don't think that will clash with anything.

rolfbjarne commented on src/SwiftRuntimeLibrary/src/SwiftRuntimeLibrary.csproj:
```diff
@@ -0,0 +1,16 @@
+    <OutputType>Library</OutputType>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>disable</Nullable>
```
Why not nullable?

rolfbjarne commented on src/SwiftRuntimeLibrary/tests/SwiftRuntimeLibrary.Tests.csproj:
```diff
@@ -0,0 +1,16 @@
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>disable</Nullable>
```
Same

rolfbjarne commented on src/SwiftRuntimeLibrary/src/SwiftCore.xml:
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="utf-8"?>
+<xamtypedatabase version="1.0">
```
```suggestion
<swifttypedatabase version="1.0">
```

rolfbjarne commented on src/SwiftRuntimeLibrary/src/SwiftCore.xml:
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="utf-8"?>
+<xamtypedatabase version="1.0">
+    <entities>
+        <entity sharpNameSpace="System" sharpTypeName="byte" entityType="Scalar" blitable="true" size="1" stride="1">
```
Technically we're creating managed bindings, not C# bindings, so use `managed` instead of `sharp` in the attribute names:

```suggestion
        <entity managedNamespace="System" managedTypeName="byte" entityType="Scalar" blitable="true" size="1" stride="1">
```

rolfbjarne commented on src/SwiftRuntimeLibrary/src/SwiftCore.xml:
```diff
@@ -0,0 +1,50 @@
+        <entity sharpNameSpace="System" sharpTypeName="byte" entityType="Scalar" blitable="true" size="1" stride="1">
+            <typedeclaration kind="struct" name="Swift.UInt8" module="Swift" />
+        </entity>
+        <entity sharpNameSpace="System" sharpTypeName="sbyte" entityType="Scalar" blitable="true" size="1" stride="1">
```
`blittable` instead of `blitable` matches how we write it everywhere else:

```suggestion
        <entity sharpNameSpace="System" sharpTypeName="sbyte" entityType="Scalar" blittable="true" size="1" stride="1">
```

rolfbjarne commented on src/SwiftRuntimeLibrary/src/SwiftCore.xml:
```diff
@@ -0,0 +1,50 @@
+        <entity sharpNameSpace="System" sharpTypeName="sbyte" entityType="Scalar" blitable="true" size="1" stride="1">
+            <typedeclaration kind="struct" name="Swift.Int8" module="Swift" />
+        </entity>
+        <entity sharpNameSpace="System" sharpTypeName="short" entityType="Scalar" blitable="true" size="2" stride="2">
```
Is `sharpTypeName` the C# keyword for the type? If so, IMHO it would make more sense to use the BCL type name instead (otherwise there's a weird mix of C#-isms here):

```suggestion
        <entity sharpNameSpace="System" sharpTypeName="Int16" entityType="Scalar" blitable="true" size="2" stride="2">
```

Then somewhere else there can be a mapping between BCL types and the corresponding built-in C# type if so desired.

kotlarmilos replied:
Thanks, updated.

kotlarmilos committed:
Use BCL managed types and enable nullable reference types

AaronRobinsonMSFT commented on src/SwiftRuntimeLibrary/src/ISwiftObject.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace Swift.Runtime
+{
+    public unsafe interface ISwiftObject : IDisposable
```
I suggest getting into the [habit of documenting public APIs](https://learn.microsoft.com/dotnet/csharp/language-reference/xmldoc/) during development. It helps with reviews and eventual public documentation efforts.

rolfbjarne replied:
If you set `GenerateDocumentationFile=true` in the csproj, you'll get warnings if you have undocumented APIs. Then setting `TreatWarningsAsErrors=true` and you'll never forget to document anyhing.

kotlarmilos replied:
Updated, thanks.

AaronRobinsonMSFT commented on src/SwiftRuntimeLibrary/src/SwiftCore.xml:
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="utf-8"?>
+<swifttypedatabase version="1.0">
+    <entities>
+        <entity managedNameSpace="System" managedTypeName="Byte" entityType="Scalar" blittable="true" size="1" stride="1">
```
Is this schema documented somewhere?

kotlarmilos replied:
I don't believe the schema is standardized. I've updated the docs to document the schema.

kotlarmilos committed:
Add Swift type database structure to README.md and add ISwiftObject interface documentation

kotlarmilos committed:
Add public access modifier to SwiftObject

jkotas commented on docs/README.md:
```diff
@@ -52,9 +52,40 @@ The table below lists the Swift types and their corresponding C# types.
+        - `managedTypeName`: Specifies the managed type name of the data type.
+        - `entityType`: Specifies the type of entity.
+        - `blittable`: Specifies whether the data type is blittable (`true` or `false`).
+        - `size`: Specifies the size of the data type in bytes.
```
Does this mean that this database file is architecture specific? Or are sizes of all types that can show up here assumed to be architecture neutral?

kotlarmilos replied:
I wouldn't go with architecture-specific database if possible. At this stage, we don't use these fields and they can be excluded.

kotlarmilos replied:
The `entityType`, `blittable`, `size`, and `stride` attributes have been removed for simplicity.

AaronRobinsonMSFT commented on src/SwiftRuntimeLibrary/src/SwiftCore.xml:
```diff
@@ -0,0 +1,50 @@
+        <entity managedNameSpace="System" managedTypeName="bool" entityType="Scalar" blittable="false" size="4" stride="4">
+            <typedeclaration kind="struct" name="Swift.Bool" module="Swift" />
+        </entity>
+        <entity managedNameSpace="SwiftRuntimeLibrary" managedTypeName="UnsafeRawPointer" entityType="Scalar" blittable="false" size="0" stride="0">
+            <typedeclaration kind="struct" name="Swift.UnsafeRawPointer" module="Swift" />
+        </entity>
+        <entity managedNameSpace="SwiftRuntimeLibrary" managedTypeName="UnsafeMutableRawPointer" entityType="Scalar" blittable="false" size="0" stride="0">
```
Can Swift have zero sized data types? If so, are they exposed in the stable ABI? I don't recall standard C or C++ permitting zero sized data types, but could be mistaken. Either way, are these really size 0?

kotlarmilos replied:
Look at the previous comment, I suggest excluding it until it becomes necessary.

stephen-hawley replied:
Just as an FYI, Swift supports 0 length value types.

AaronRobinsonMSFT commented on src/SwiftRuntimeLibrary/src/SwiftCore.xml:
```diff
@@ -0,0 +1,50 @@
+        <entity managedNameSpace="System" managedTypeName="UIntPtr" entityType="Scalar" blittable="true" size="8" stride="8">
+            <typedeclaration kind="struct" name="Swift.UInt" module="Swift" />
+        </entity>
+        <entity managedNameSpace="System" managedTypeName="bool" entityType="Scalar" blittable="false" size="4" stride="4">
```
Is this for .NET, Swift, or the marshaled type? Both .NET and Swift have `bool` as 1 byte - `MemoryLayout<Bool>.size`. The built in marshaller defaults to the `BOOL` size, which is 4. However, if we are documenting that here then I don't understand this file's purpose. Are these for .NET, Swift or interop types? What are we doing about [`DisableDefaultMarshalling`](https://learn.microsoft.com/dotnet/api/system.runtime.compilerservices.disableruntimemarshallingattribute), since in that case the marshalled `bool` would indeed be 1.

stephen-hawley replied:
This file is meant for mapping swift types to .NET types. In a quick look-over of the code, size, stride and blittable are artifacts and can go away entirely.
In the dim dark ages of BTfS, I tried to create two classes of struct: blittable (and fully representable in C# with no wrappers) and non-blittable. This was an endless source of bugs and I removed it.

kotlarmilos committed:
Remove type-specific information from the type database

kotlarmilos committed:
Add bindings generator with a simple parser and emitter

The bindings generator includes a parser, marshaller, and emitter. The parser generates module declarations by consuming an ABI JSON file. The marshaller facilitates type marshalling between C# and Swift, while the emitter is a string-based IndentedTextWriter that outputs C# source code.

kotlarmilos committed:
Update docs and add HelloWorld sample

kotlarmilos commented:
This PR has been updated with a bindings generator that includes a simple parser and emitter (refer to the updated description and docs). Please note that this is an MVP for CryptoKit dev templates.

rolfbjarne commented on docs/README.md:
````diff
@@ -105,38 +132,29 @@ In the example, the user's code references the `HelloLibraryBindings` namespace
-    }
-}
-```
+The aggregation of the public ABI is done through the `ISwiftParser` interface. This interface defines the layout for concrete implementations responsible for parsing and collecting ABI information. Two implementations exist: Swift ABI parser and Swift interface parser. The Swift ABI parser aggregates ABI information based on an ABI json file. The Swift interface parser is designed to handle `.sw ...
````
```suggestion
The aggregation of the public ABI is done through the `ISwiftParser` interface. This interface defines the layout for concrete implementations responsible for parsing and collecting ABI information. Two implementations exist: Swift ABI parser and Swift interface parser. The Swift ABI parser aggregates ABI information based on an ABI json file. The Swift interface parser is designed to handle `.swiftinterface` files. The `.swiftinterface` file doesn't contain mangled names, and the parser should consume the dynamic library (`.dylib`) to generate declarations. Currently, the tooling only implements the Swift ABI parser.
```

rolfbjarne commented on src/SwiftRuntimeLibrary/src/TypeDatabase.cs:
```diff
@@ -0,0 +1,116 @@
+
+using System.Xml;
+
+namespace SwiftRuntimeLibrary
```
```suggestion
namespace Swift.Runtime
```

kotlarmilos replied:
Do you think we should change the .csproj name to `SwiftRuntime.csproj`/`SwiftRuntime.Tests.csproj`?

rolfbjarne replied:
Maybe Swift.Runtime[.Test].csproj?

kotlarmilos replied:
Changed to Swift.Bindings and Swift.Runtime for clarity.

rolfbjarne commented on src/SwiftRuntimeLibrary/src/TypeDatabase.cs:
```diff
@@ -0,0 +1,116 @@
+        {
+            XmlNode? entitiesNode = xmlDoc.SelectSingleNode("//swifttypedatabase/entities");
+
+            if (entitiesNode != null)
+            {
+                foreach (XmlNode? entityNode in entitiesNode.ChildNodes)
+                {
+                    XmlNode? typeDeclarationNode = entityNode?.SelectSingleNode("typedeclaration");
+                    if (typeDeclarationNode == null)
+                        throw new Exception("Invalid XML structure: 'typedeclaration' node not found.");
+                    
+                    string? swiftName = typeDeclarationNode?.Attributes?["name"]?.Value;
+                    string? csharpName = entityNode?.Attributes?["managedTypeName"]?.Value;
+                    if (swiftName == null || csharpName == null)
+                        throw new Exception("Invalid XML structure: Missing attributes.");
+
+                    _swiftToCSharpMapping.Add(swiftName, csharpName);
+                }
+            }
+            else
+            {
+                throw new Exception("Invalid XML structure: 'entities' node not found.");
+            }
+        }
```
This can be turned around to reduce indentation:

```suggestion
            if (entitiesNode == null)
            	throw new Exception("Invalid XML structure: 'entities' node not found.");

            foreach (XmlNode? entityNode in entitiesNode.ChildNodes)
            {
                XmlNode? typeDeclarationNode = entityNode?.SelectSingleNode("typedeclaration");
                if (typeDeclarationNode == null)
                    throw new Exception("Invalid XML structure: 'typedeclaration' node not found.");
                    
                string? swiftName = typeDeclarationNode?.Attributes?["name"]?.Value;
                string? csharpName = entityNode?.Attributes?["managedTypeName"]?.Value;
                if (swiftName == null || csharpName == null)
                    throw new Exception("Invalid XML structure: Missing attributes.");

                _swiftToCSharpMapping.Add(swiftName, csharpName);
            }
        }
```

rolfbjarne commented on src/SwiftRuntimeLibrary/src/TypeDatabase.cs:
```diff
@@ -0,0 +1,116 @@
+                    ReadVersion1_0(xmlDoc);
+                    break;
+                default:
+                    throw new Exception("Unsupported XML version.");
```
```suggestion
                    throw new Exception("Unsupported database version {0} in {1}.", version, file);
```

rolfbjarne commented on src/SwiftRuntimeLibrary/src/TypeDatabase.cs:
```diff
@@ -0,0 +1,116 @@
+            if (!ValidateXmlSchema(xmlDoc))
+                throw new Exception("Invalid XML schema.");
+
+            switch (xmlDoc.DocumentElement?.Attributes?["version"]?.Value)
```
```suggestion
            var version = xmlDoc.DocumentElement?.Attributes?["version"]?.Value;
            switch (version)
```

rolfbjarne commented on src/SwiftRuntimeLibrary/src/TypeDatabase.cs:
```diff
@@ -0,0 +1,116 @@
+            XmlDocument xmlDoc = new XmlDocument();
+            xmlDoc.Load(file);
+            if (!ValidateXmlSchema(xmlDoc))
+                throw new Exception("Invalid XML schema.");
```
```suggestion
                throw new Exception("Invalid XML schema in {0}.", file);
```

kotlarmilos committed:
Update docs/README.md

Co-authored-by: Rolf Bjarne Kvinge <rolf@xamarin.com>

kotlarmilos committed:
Update src/SwiftRuntimeLibrary/src/TypeDatabase.cs

Co-authored-by: Rolf Bjarne Kvinge <rolf@xamarin.com>

kotlarmilos committed:
Update src/SwiftRuntimeLibrary/src/TypeDatabase.cs

Co-authored-by: Rolf Bjarne Kvinge <rolf@xamarin.com>

kotlarmilos committed:
Update src/SwiftRuntimeLibrary/src/TypeDatabase.cs

Co-authored-by: Rolf Bjarne Kvinge <rolf@xamarin.com>

kotlarmilos committed:
Update src/SwiftRuntimeLibrary/src/TypeDatabase.cs

Co-authored-by: Rolf Bjarne Kvinge <rolf@xamarin.com>

kotlarmilos committed:
Fix exception message formatting

kotlarmilos committed:
Refactor Swift.Runtime namespace

kotlarmilos committed:
Refactor project naming: Swift.Runtime and Swift.Bindings

kotlarmilos committed:
Use deinit calls for Swift projections in C# by default

kotlarmilos committed:
Add support for UnsafeRawPointer

This commit introduces support for the UnsafeRawPointer type. It also adds a test case for the existing CryptoKit bindings.

kotlarmilos committed:
Merge branch 'feature/swift-bindings' of github.com:dotnet/runtimelab into swift-bindings/runtime-library

kotlarmilos committed:
Update dotnet version

matouskozak commented on src/Swift.Runtime/src/TypeDatabase.cs:
```diff
@@ -0,0 +1,123 @@
+        /// <returns><c>true</c> if the Swift type name maps to a C# value type; otherwise <c>false</c>.</returns>
+        public bool IsValueType(string swiftName)
+        {
+            return _swiftToCSharpMapping.TryGetValue(swiftName, out string? csharpName);
```
Do we expect the mapping in `src/Swift.Runtime/src/TypeDatabase.xml` to grow in the future so that it would contain also other types that don't match to C# value types? For example, strings?

kotlarmilos replied:
The `TypeDatabase.xml` should provide mappings for common C# types. It can be expanded in the future, and we need to project each type pattern individually. For non-trivial mappings, a projection will be created within the `Swift.Runtime` namespace (i.e. unsafe buffer pointers). Currently, the type database attempts to find a known mapping. If not found, the type is then searched for within `Swift.Runtime` namespace:
```csharp
/// <summary>
/// Gets the C# type name corresponding to the specified Swift type name. 
/// The method first tries to find a known mapping, and if that fails, it looks for a type in Swift.Runtime.
/// </summary>
/// <param name="swiftTypeName">The Swift type name.</param>
/// <returns>The corresponding C# type name.</returns>
public string[] GetCSharpTypeName(string swiftTypeName)
{
    // Try to find a known mapping
    if (_swiftToCSharpMapping.TryGetValue(swiftTypeName, out string? csharpTypeName))
    {
        return new string [] { "System", csharpTypeName };
    }

    // Try to find a type in Swift.Runtime
    Type? swiftRuntimeType = Type.GetType($"Swift.Runtime.{swiftTypeName}");
    if (swiftRuntimeType != null)
        return new string [] { "Swift.Runtime", swiftRuntimeType.Name };
    
    throw new Exception($"No mapping for {swiftTypeName} type found.");
}
```

matouskozak replied:
Thank you for the explanation

matouskozak commented on src/Swift.Bindings/src/Model/MethodDecl.cs:
```diff
@@ -0,0 +1,26 @@
+        /// <summary>
+        /// Signature of the method.
+        /// </summary>
+        public required IEnumerable<TypeDecl> Signature { get; set; }
```
I wonder if having signature represented just by a simple `IEnumerable<TypeDecl>` will be enough in the future or if it would be better to declare `SignatureDecl` and split return and params into separate fields. 

My concerns here is the possibility of implicit `self`/`this` but I'm unsure if that will concern us in the future. Nonetheless, for the prototype we can keep this for simplicity.

kotlarmilos replied:
It's a good idea, and we can consider implementing it in the future. Swift ABI signature is presented in this way (return type + params).

matouskozak commented on src/Swift.Bindings/src/Parser/BindingsGenerator.SwiftABIParser.cs:
```diff
@@ -0,0 +1,130 @@
+    /// </summary>
+    public sealed class RootNode
+    {
+        public required string Kind { get; set; }
```
Maybe enum would be a better long term type for `Kind` but I suppose we will move away from parsing Swift ABI so it is probably okay to use string for now.

kotlarmilos replied:
If `Kind` was implemented as enum, it would impact the JSON deserialization process, expecting `int` type instead of `string`:
```csharp
var abiRoot = JsonConvert.DeserializeObject<ABIRootNode>(jsonContent);
```

> will move away from parsing Swift ABI so it is probably okay to use string for now.

I think it is a good idea to wait until parser becomes more stable.

matouskozak commented on src/Swift.Runtime/src/ISwiftObject.cs:
Is the ISwiftObject used for something? I couldn't find its usage in the PR.

kotlarmilos replied:
Not needed, removed for now.

matouskozak reviewed (approved):
Good job! Looking good and much easier to read through than the original big PR. I put some comments but they are more questions about the future so no need to address them in this PR.

kotlarmilos committed:
Add support for UnsafeBufferPointer types

This commit introduces support for the UnsafeBufferPointer types. It also adds a test case for the CryptoKit bindings.

kotlarmilos committed:
Update Swift ABI parser to capture dependencies and C# mapped types

kotlarmilos committed:
Update docs with UnsafeBufferPointer section

kotlarmilos commented:
This PR includes projection tooling skeleton and implements CryptoKit dev templates with unsafe (buffer) pointers. We will move forward with surfacing https://developer.apple.com/documentation/cryptokit/chachapoly enum and https://developer.apple.com/documentation/cryptokit/symmetrickey struct in a subsequent PR.

Please review this PR and provide feedback.

AaronRobinsonMSFT commented on .gitignore:
```diff
@@ -190,3 +190,15 @@ Session.vim
 # VS debug support files
 launchSettings.json
+
+# Swift output files
+*.swiftinterface*
+*.swiftdoc*
+*.swiftmodule*
+*.abi*
+*.swiftsourceinfo*
+*.dylib
+
+# Testing artifacts
+testing/
+src/samples/**/*Bindings.cs
```
These should all go under `artifacts/`. In the current experiment form, this seems okay, but prior to merging into dotnet/runtime, all outputs should be under `artifacts/`.

AaronRobinsonMSFT replied:
Oops. Sorry, I thought we were working in a dotnet/runtime fork. Never mind.

AaronRobinsonMSFT commented on docs/functional-outline.png:
This would ideally be an SVG instead of PNG.

kotlarmilos replied:
Updated.

AaronRobinsonMSFT commented on src/Swift.Bindings/src/Model/MethodDecl.cs:
```diff
@@ -0,0 +1,32 @@
+    /// <summary>
+    /// Represents a method declaration.
+    /// </summary>
+    public sealed record MethodDecl
```
Using `record`s in public APIs hasn't been confirmed as yet. This is likely okay for the current state. Just be aware that if this is going to be publicly consumable, the API review team may push back.

AaronRobinsonMSFT reviewed (approved):
LGTM

jkotas commented on docs/README.md:
```diff
@@ -44,6 +44,8 @@ The table below lists the Swift types and their corresponding C# types.
 | `Swift.UInt16`                  | `ushort` |
 | `Swift.Int8`                    | `sbyte`  |
 | `Swift.UInt8`                   | `byte`   |
+| `Swift.UnsafePointer`           | `void*`  |
```
I know we touched on this before - it does not feel right to drop the type information for the strongly typed pointers. Strong typing is an extra level of protection against programming bugs that we are giving away here.

kotlarmilos replied:
Ok, as we discussed, a thin layer would be useful here.

kotlarmilos replied:
Types for unsafe pointers have been added. Mutable pointers are projected as structs with generics in C# to address methods overload issues and enable future extensibility.

jkotas commented on src/Swift.Runtime/src/Library/UnsafeBufferPointer.cs:
```diff
@@ -0,0 +1,72 @@
+    public unsafe readonly struct UnsafeBufferPointer
+    {
+        private readonly void* _Position;
+        public readonly nint Count;
```
Swift does not seem to have fields as first-class concept. Their docs only talk about properties: https://docs.swift.org/swift-book/documentation/the-swift-programming-language/properties/

I am wondering whether mapping stored properties to fields in C# is conceptually correct.

Is it a breaking change in Swift to change a stored property to a computed property in non-frozen type? I would expect that it is not a breaking change - can you confirm that?

If computed properties and stored properties are interchangeable in Swift, we should map both to properties in C#. Ie make this `public nint Count { get; };` or `private nint _count; public nint Count => _count;`.

kotlarmilos replied:
Good catch. This is not a breaking change in Swift. I confirmed it by consuming a library with struct that evolved from
```swift
public struct MyType {
    public var myProperty: Int = 0
    public init() {}
}
```
to
```swift
public struct MyType {
    private var _myProperty: Int = 0
    public var myProperty: Int {
        get { _myProperty }
        set { _myProperty = newValue }
    }

    public init() {}
}
```

kotlarmilos replied:
Buffer pointers are updated with `private readonly nint _count; public nint Count => _count;`.

stephen-hawley replied:
IIRC, if a property is stored, you still get getters and setter generated by the compiler. Since BTfS doesn't deal with struct lowering, it maps properties (stored or otherwise) as C# properties.

kotlarmilos committed:
Implement mutable pointers as generics in C#

This commit implements mutable pointers as generics in C# and update CryptoKit test cases. The parser is updated to process Swift generic types and projects them into C# generic types.

kotlarmilos committed:
Add P/Invoke bool test case

jkotas commented on src/Swift.Runtime/src/Library/UnsafeBufferPointer.cs:
```diff
@@ -0,0 +1,73 @@
+            _count = count;
+        }
+
+        public void* BaseAddress => _position;
```
```suggestion
        public T* BaseAddress => _position;
```

jkotas commented on src/Swift.Runtime/src/Library/UnsafeBufferPointer.cs:
```diff
@@ -0,0 +1,73 @@
+            _count = count;
+        }
+
+        public void* BaseAddress => _position;
```
```suggestion
        public T* BaseAddress => _position;
```

jkotas replied:
All `void*` in the strongly typed wrappers should be `T*`.

jkotas commented on src/Swift.Runtime/src/Library/UnsafePointer.cs:
```diff
@@ -0,0 +1,61 @@
+    public unsafe readonly struct UnsafePointer<T>
+    {
+        private readonly void* _rawValue;
+        public UnsafePointer(void* _rawValue)
```
The argument name should not have underscore. Also, should the argument name match the property name that it corresponds to `UnsafePointer(void* poin_rawValue)`?

jkotas commented on src/Swift.Runtime/src/Library/UnsafePointer.cs:
```diff
@@ -0,0 +1,61 @@
+    // <summary>
+    // Represents Swift UnsafePointer in C#.
+    // </summary>
+    public unsafe readonly struct UnsafePointer<T>
```
I think it would make sense to have implicit conversions to/from T* for these types.

jkotas replied:
(And to/from `void*` for the non-generic variant.)

jkotas commented on src/Swift.Runtime/src/Library/UnsafeBufferPointer.cs:
```diff
@@ -0,0 +1,73 @@
+    {
+        private readonly void* _position;
+        private readonly void* _end;
+        public UnsafeRawBufferPointer(void* start, nint count)
```
```suggestion
        public UnsafeRawBufferPointer(void* baseAddress, nint count)
```
Match property name?

kotlarmilos committed:
Update pointers with implicit operators
